In [1]:
!pip install surprise
from surprise import Dataset, accuracy
from surprise.model_selection import cross_validate, GridSearchCV
from surprise.prediction_algorithms import BaselineOnly, SVD, SVDpp, NMF

### Завантажуємо movielens-100k датасет(алгоритм запропонує завантажити його, якщо необхідно)

In [2]:
data = Dataset.load_builtin('ml-100k')

Dataset ml-100k could not be found. Do you want to download it? [Y/n] y
Trying to download dataset from https://files.grouplens.org/datasets/movielens/ml-100k.zip...
Done! Dataset ml-100k has been saved to /Users/maksm/.surprise_data/ml-100k


### Крос-валідація на моделі матричної факторизації SVD

In [4]:
algorithm_svd = SVD()

cross_validate(algorithm_svd, data, measures=["RMSE", "MAE"], cv=5, verbose=True) 

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9419  0.9340  0.9356  0.9400  0.9339  0.9371  0.0033  
MAE (testset)     0.7425  0.7373  0.7368  0.7405  0.7364  0.7387  0.0024  
Fit time          0.75    0.70    0.79    0.77    0.76    0.75    0.03    
Test time         0.13    0.13    0.13    0.07    0.13    0.12    0.02    


{'test_rmse': array([0.94193711, 0.93401087, 0.93562071, 0.9400364 , 0.93389946]),
 'test_mae': array([0.74249592, 0.73725943, 0.73679672, 0.74051137, 0.73636439]),
 'fit_time': (0.7466118335723877,
  0.7007889747619629,
  0.7913739681243896,
  0.7676379680633545,
  0.7615222930908203),
 'test_time': (0.1345381736755371,
  0.12739300727844238,
  0.1257481575012207,
  0.07040786743164062,
  0.13417983055114746)}

### Результати крос-валідації для алгоритму SVD показують наступне:

* RMSE (середньо-квадратична помилка): Середнє значення: 0.9371 Стандартне відхилення: 0.0033

* MAE (середня абсолютна помилка): Середнє значення: 0.7387 Стандартне відхилення: 0.0024

* Час навчання (Fit time): Середнє значення: 0.75 сек Стандартне відхилення: 0.03 сек

* Час тестування (Test time): Середнє значення: 0.12 сек Стандартне відхилення: 0.02 сек

Ці результати вказують на те, що модель SVD має досить низьке значення середньо-квадратичної та середньої абсолютної помилок, що вказує на гарну точність моделі при прогнозуванні рейтингів. Також, час навчання та тестування є прийнятними і не сильно відрізняються між підсумковими фолдами.

### Крос-валідація на моделі матричної факторизації SVD++

In [5]:
algorithm_svdpp = SVDpp()

cross_validate(algorithm_svdpp, data, measures=["RMSE", "MAE"], cv=5, verbose=True) 

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9188  0.9223  0.9155  0.9200  0.9198  0.9193  0.0022  
MAE (testset)     0.7199  0.7216  0.7190  0.7218  0.7228  0.7210  0.0014  
Fit time          11.44   11.58   11.46   11.33   11.61   11.48   0.10    
Test time         3.19    3.12    3.29    3.30    2.92    3.16    0.14    


{'test_rmse': array([0.91881797, 0.92231464, 0.91547459, 0.92002262, 0.91976026]),
 'test_mae': array([0.71988952, 0.72156595, 0.71897951, 0.72178284, 0.72278958]),
 'fit_time': (11.440852165222168,
  11.58205795288086,
  11.455650806427002,
  11.328891277313232,
  11.60959792137146),
 'test_time': (3.188601016998291,
  3.116645097732544,
  3.294145107269287,
  3.3007419109344482,
  2.922583818435669)}

### Результати крос-валідації для алгоритму SVDpp показують наступне:

* RMSE (середньо-квадратична помилка): Середнє значення: 0.9193 Стандартне відхилення: 0.0022

* MAE (середня абсолютна помилка): Середнє значення: 0.7210 Стандартне відхилення: 0.0014

* Час навчання (Fit time): Середнє значення: 11.48 сек Стандартне відхилення: 0.10 сек

* Час тестування (Test time): Середнє значення: 3.16 сек Стандартне відхилення: 0.14 сек

Алгоритм SVDpp показує гарні результати з точки зору точності прогнозів, яка вимірюється RMSE та MAE. Відзначимо, що SVDpp має більший час навчання та тестування порівняно з SVD, оскільки SVDpp включає додаткові параметри та фактори, що призводить до більшої обчислювальної складності. Важливо збалансувати точність та часові витрати в залежності від конкретних вимог нашого застосування.

### Крос-валідація на моделі матричної факторизації NMF

In [6]:
algorithm_nmf = NMF()

cross_validate(algorithm_nmf, data, measures=["RMSE", "MAE"], cv=5, verbose=True) 

Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9664  0.9551  0.9638  0.9677  0.9600  0.9626  0.0046  
MAE (testset)     0.7582  0.7545  0.7553  0.7586  0.7542  0.7562  0.0019  
Fit time          1.44    1.40    1.29    1.33    1.46    1.38    0.07    
Test time         0.14    0.13    0.07    0.13    0.13    0.12    0.03    


{'test_rmse': array([0.96637839, 0.95508317, 0.96380021, 0.967743  , 0.95997816]),
 'test_mae': array([0.75819357, 0.75450874, 0.75530731, 0.75861325, 0.75416907]),
 'fit_time': (1.4414598941802979,
  1.4037728309631348,
  1.288954734802246,
  1.3257789611816406,
  1.463810920715332),
 'test_time': (0.13592910766601562,
  0.12946510314941406,
  0.06757807731628418,
  0.13198399543762207,
  0.13120007514953613)}

### Результати крос-валідації для алгоритму NMF (Non-Negative Matrix Factorization) показують наступне:

* RMSE (середньо-квадратична помилка): Середнє значення: 0.9626 Стандартне відхилення: 0.0046

* MAE (середня абсолютна помилка): Середнє значення: 0.7562 Стандартне відхилення: 0.0019

* Час навчання (Fit time): Середнє значення: 1.38 сек Стандартне відхилення: 0.07 сек

* Час тестування (Test time): Середнє значення: 0.12 сек Стандартне відхилення: 0.03 сек

Алгоритм NMF також представляє певний рівень точності, але його значення RMSE та MAE трошки вищі порівняно з SVD та SVDpp. Важливо врахувати, що NMF використовує обмеження невід'ємності, що може бути корисним в деяких випадках, особливо якщо значення рейтингів не може бути від'ємним. Зате, час навчання та тестування для NMF є прийнятними.

### Функція для підбору найкращих параметрів за допомогою крос-валідації

In [12]:
def model_evaluation(algo, param_grid):
    print(f"\nAlgorithm: {algo.__name__}")
    
    # Створення об'єкта GridSearchCV
    grid_search = GridSearchCV(algo, param_grid, measures=['rmse', 'mae'], cv=3)
    
    # Запуск крос-валідації
    grid_search.fit(data)
    
    # Виведення найкращих параметрів та результатів
    print(f"Best parameters: {grid_search.best_params}")
    print(f"Best RMSE: {grid_search.best_score['rmse']:.4f}")

    # Оцінка моделі на весь набір даних (тренувальний та тестовий)
    algo_instance = grid_search.best_estimator['rmse']
    algo_instance.fit(data.build_full_trainset())
    predictions = algo_instance.test(data.build_full_trainset().build_testset())
    
    # Оцінка моделі
    print(f"\nTrainset Accuracy (RMSE): {accuracy.rmse(predictions):.4f}")

### Визначення найкращих параметрів моделі матричної факторизації SVD за допомогою крос-валідації

In [15]:
param_grid = {
    'n_factors': [10, 20, 30],
    'n_epochs': [1, 5, 10],
    'lr_all': [0.002, 0.005],
    'reg_all': [0.02, 0.1]
}

algorithm_svd = SVD
model_evaluation(algorithm_svd, param_grid)


Algorithm: SVD
Best parameters: {'rmse': {'n_factors': 10, 'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.02}, 'mae': {'n_factors': 10, 'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.02}}
Best RMSE: 0.9483
RMSE: 0.9139

Trainset Accuracy (RMSE): 0.9139


#### Аналіз результатів:

* Після пошуку оптимальних параметрів за допомогою крос-валідації найкращі параметри для алгоритму SVD визначені як:

n_factors: 10 n_epochs: 10 lr_all: 0.005 reg_all: 0.02

* Найкращий RMSE: 0.9483

Найкращі параметри були використані для тренування моделі на повному тренувальному наборі даних, і її точність була оцінена на всьому наборі даних. Оцінка моделі на весь набір даних показала RMSE рівний 0.9139.

Загалом, результати показують, що модель SVD показала гарні характеристики. Найкращі параметри можуть бути змінені під час налаштування моделі. Значення RMSE є досить низькими, що свідчить про добру точність моделі у прогнозуванні рейтингів на всьому наборі даних. Таким чином, SVD із відповідними параметрами є ефективною моделлю для цього датасету, і вона може бути використана для рекомендаційного аналізу з задовільною точністю.

### Визначення найкращих параметрів моделі матричної факторизації SVDpp за допомогою крос-валідації

In [16]:
param_grid_svdpp = {
    'n_factors': [10, 20, 30],
    'n_epochs': [1, 5, 10],
    'lr_all': [0.002, 0.005],
    'reg_all': [0.02, 0.1]
}

algorithm_svdpp = SVDpp
model_evaluation(algorithm_svdpp, param_grid)


Algorithm: SVDpp
Best parameters: {'rmse': {'n_factors': 30, 'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.02}, 'mae': {'n_factors': 30, 'n_epochs': 10, 'lr_all': 0.005, 'reg_all': 0.02}}
Best RMSE: 0.9425
RMSE: 0.8730

Trainset Accuracy (RMSE): 0.8730


#### Аналіз результатів:

* Після пошуку оптимальних параметрів за допомогою крос-валідації найкращі параметри для алгоритму SVD визначені як:

n_factors: 30 n_epochs: 10 lr_all: 0.005 reg_all: 0.02

* Найкращий RMSE: 0.9425

Найкращі параметри були використані для тренування моделі на повному тренувальному наборі даних, і її точність була оцінена на всьому наборі даних. Оцінка моделі на весь набір даних показала RMSE рівний 0.8730. 

Загалом, результати показують, що модель SVD++ показала гарні характеристики. Найкращі параметри можуть бути змінені під час налаштування моделі. Значення RMSE є досить низькими, що свідчить про добру точність моделі у прогнозуванні рейтингів на всьому наборі даних. Таким чином, SVD++ із відповідними параметрами також є ефективною моделлю для цього датасету, і вона може бути використана для рекомендаційного аналізу з задовільною точністю.

### Визначення найкращих параметрів моделі матричної факторизації NMF за допомогою крос-валідації

In [18]:
param_grid = {'n_factors': [10, 20, 30],
              'n_epochs': [1, 5, 10],
              'reg_pu': [0.06, 0.08, 0.1],
              'reg_qi': [0.06, 0.08, 0.1],
              'reg_bu': [0.02, 0.04, 0.6],
              'reg_bi': [0.02, 0.04, 0.6]}

algorithm_nmf = NMF
model_evaluation(algorithm_nmf, param_grid)


Algorithm: NMF
Best parameters: {'rmse': {'n_factors': 10, 'n_epochs': 5, 'reg_pu': 0.1, 'reg_qi': 0.1, 'reg_bu': 0.04, 'reg_bi': 0.6}, 'mae': {'n_factors': 10, 'n_epochs': 5, 'reg_pu': 0.1, 'reg_qi': 0.1, 'reg_bu': 0.04, 'reg_bi': 0.6}}
Best RMSE: 1.0775
RMSE: 1.0153

Trainset Accuracy (RMSE): 1.0153


### Аналіз результатів:

За результатами пошуку оптимальних параметрів за допомогою крос-валідації найкращі параметри для алгоритму NMF визначені як: 
n_factors: 10 n_epochs: 5 reg_pu: 0.1 reg_qi: 0.1 reg_bu: 0.04 reg_bi: 0.6

Найкращий RMSE, отриманий за допомогою крос-валідації, становить 1.0775. Після отримання найкращих параметрів модель була навчена на повному тренувальному наборі даних, і її точність була оцінена на всьому наборі даних. Оцінка моделі на весь набір даних показала RMSE рівний 1.0153.

### Висновки:

Отримані параметри та оцінки демонструють, що алгоритм NMF з обраною сіткою параметрів має гарні характеристики на даному наборі даних. Значення RMSE 1.0153 вказує на те, що середня квадратична помилка прогнозування рейтингів на весь набір даних досить низька, що може вказувати на ефективність моделі в прогнозуванні рейтингів. Загалом, отримані результати свідчать про те, що модель NMF з вибраними параметрами є ефективною для задачі рекомендаційного аналізу на даному датасеті.